### ISSUE TITLE
Push down count distinct on partition column for subqueries

### ISSUE THREAD

**Issue by marcocitus** [citusdata/citus#1877]
We optimise `count(distinct distribution_column)` queries by pushing down the `count(distinct distribution_column)` to the workers. 

```sql
SELECT x, count(distinct x) FROM test GROUP BY x;

-- worker query:
SELECT x, count(DISTINCT x) AS count FROM test_102200 test WHERE true GROUP BY x
```

However, we do not do this for subqueries, e.g.:

```sql
SELECT x, count(distinct x) FROM (SELECT test.x FROM test JOIN ref USING (y) GROUP BY test.x) sample GROUP BY x;

-- worker query:
SELECT worker_column_1 AS x, worker_column_1 AS count FROM (SELECT sample.x AS worker_column_1 FROM (SELECT test.x FROM (public.test_102200 test JOIN public.ref_102297 ref USING (y)) GROUP BY test.x) sample) worker_subquery WHERE true GROUP BY worker_column_1, worker_column_1
```

This is because in `GroupedByDisjointPartitionColumn` we do not consider subqueries. We may want to call `TargetListOnPartitionColumn` to determine whether the distinct is on a partition column of the distributed tables in the subquery.

**Comment by mtuncer** [comment]
We have a similar situation with window functions and limit pushdown.
Limit in following query should be pushed down.  However, due to processing of the query with subquery planner since it has a window function; we are unable to detect that the group by is on partition column, and there is an order by on aggregate. Therefore limit is not pushed down

```sql
SELECT
	DISTINCT ON (l_orderkey, RANK() OVER (partition by l_orderkey)) l_orderkey, count(*), RANK() OVER (partition by l_orderkey)
	FROM lineitem_hash
	GROUP BY l_orderkey
	ORDER BY l_orderkey , 3, 2
	LIMIT 5;
```

**Comment by onderkalaci** [comment]
Seems to be fixed by https://github.com/citusdata/citus/pull/4355. @marcocitus can you confirm as well? If not, please re-open the issue.

**Comment by onderkalaci** [comment]
> We have a similar situation with window functions and limit pushdown.
Limit in following query should be pushed down. However, due to processing of the query with subquery planner since it has a window function; we are unable to detect that the group by is on partition column, and there is an order by on aggregate. Therefore limit is not pushed down

same for this, fixed by #4355
